satellite/satellitedb/certdb.go

```diff
@@ -19,82 +22,96 @@ type certDB struct {
-	return nil
-}
+	defer func() {
+		if err == nil {
+			err = tx.Commit()
+		} else {
+			err = errs.Combine(err, tx.Rollback())
+		}
+	}()
 
-func (certs *certDB) tryAddPublicKey(ctx context.Context, nodeID storj.NodeID, publicKey crypto.PublicKey) (err error) {
-	defer mon.Task()(&ctx)(&err)
-	// no rows err, so create/insert an entry
-	pubbytes, err := pkcrypto.PublicKeyToPKIX(publicKey)
-	if err != nil {
-		return Error.Wrap(err)
+	if pi == nil {
+		return Error.New("Peer Identity cannot be nil")
 	}
+	chain := encodePeerIdentity(pi)
 
-	// TODO: use upsert here instead of create
-	_, err = certs.db.Create_CertRecord(ctx,
-		dbx.CertRecord_Publickey(pubbytes),
-		dbx.CertRecord_Id(nodeID.Bytes()),
-	)
-	// another goroutine might race to create the cert record, let's ignore that error
-	if pgutil.IsConstraintError(err) || sqliteutil.IsConstraintError(err) {
-		return nil
-	} else if err != nil {
+	var id int64
+	query := `SELECT id FROM certRecords WHERE peer_identity = ?;`
+	err = tx.QueryRow(certs.db.Rebind(query), chain).Scan(&id)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// create a new entry
+			_, err = tx.Exec(certs.db.Rebind(`INSERT INTO certRecords ( peer_identity, node_id, update_at ) VALUES ( ?, ?, ? );`), chain, nodeID.Bytes(), time.Now())
+			if err != nil {
+				return Error.Wrap(err)
+			}
+			return nil
+		}
 		return Error.Wrap(err)
 	}
 
+	// already public key exists, just return
 	return nil
 }
 
 // GetPublicKey gets the public key of uplink corresponding to uplink id
-func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ crypto.PublicKey, err error) {
+func (certs *certDB) GetPublicKey(ctx context.Context, nodeID storj.NodeID) (_ *identity.PeerIdentity, err error) {
 	defer mon.Task()(&ctx)(&err)
-	dbxInfo, err := certs.db.All_CertRecord_By_Id_OrderBy_Desc_UpdateAt(ctx, dbx.CertRecord_Id(nodeID.Bytes()))
+	dbxInfo, err := certs.db.All_CertRecord_By_NodeId_OrderBy_Desc_UpdateAt(ctx, dbx.CertRecord_NodeId(nodeID.Bytes()))
 	if err != nil {
 		return nil, err
 	}
 
+	if len(dbxInfo) == 0 {
+		return nil, Error.New("Invalid nodeID : %+v: %+v ", nodeID.String(), err)
+	}
+
 	// the first indext always holds the lastest of the keys
-	pubkey, err := pkcrypto.PublicKeyFromPKIX(dbxInfo[0].Publickey)
-	if err != nil {
-		return nil, Error.New("Failed to extract Public Key from Order: %+v", err)
+
+	peer, err := decodePeerIdentity(ctx, dbxInfo[0].PeerIdentity)
+	return peer, Error.Wrap(err)
+}
+
+func encodePeerIdentity(pi *identity.PeerIdentity) []byte {
+	var chain []byte
+	chain = append(chain, pi.Leaf.Raw...)
+	chain = append(chain, pi.CA.Raw...)
+	for _, cert := range pi.RestChain {
+		chain = append(chain, cert.Raw...)
 	}
-	return pubkey, nil
+	return chain
 }
 
-// GetPublicKeys gets the public keys of a storagenode corresponding to storagenode id
-func (certs *certDB) GetPublicKeys(ctx context.Context, nodeID storj.NodeID) (pubkeys []crypto.PublicKey, err error) {
```

Comment:
**Rational for removing `GetPublicKeys()`:**
Didn't see a scenario of needing more than one peer-identity for validating. This certRecord DB does store all previous peer-identities for a particular Storagenode, if it decides to have a new identity.